lib/rgeo/cartesian/planar_graph.rb

```diff
@@ -0,0 +1,400 @@
+          new_edges.each do |edge|
+            create_half_edge(edge)
+          end
+
+          intersection_map.each do |seg, ints|
+            compute_split_edges(seg, ints)
+          end
+
+          link_half_edges
+        end
+
+        private
+
+        # Creates a map of +proper_intersections+ for each segment
+        # from a sweepline intersector.
+        #
+        # Can be used to determine which edges need to be split
+        # after adding edges.
+        def intersection_map
+          intersector = SweeplineIntersector.new(edges)
+          intersections = intersector.proper_intersections
+
+          intersection_map = {}
+          intersections.each do |int|
+            # check the int_point against each edge.
+            # if it is not on the boundary of the edge, add it to the
+            # list of intersections for that edge.
+            s1_intersects = int.point != int.s1.s && int.point != int.s1.e
+            if s1_intersects
+              unless intersection_map[int.s1]
+                intersection_map[int.s1] = []
+              end
+              intersection_map[int.s1] << int.point
+            end
+
+            s2_intersects = int.point != int.s2.s && int.point != int.s2.e
+            next unless s2_intersects
+            unless intersection_map[int.s2]
+              intersection_map[int.s2] = []
+            end
+            intersection_map[int.s2] << int.point
+          end
+          intersection_map
+        end
+
+        def create_half_edges
+          @edges.each do |edge|
+            create_half_edge(edge)
+          end
+        end
+
+        def create_half_edge(edge)
+          e1, e2 = HalfEdge.from_edge(edge)
+          insert_half_edge(e1)
+          insert_half_edge(e2)
+        end
+
+        def insert_half_edge(he)
+          unless incident_edges[he.origin.coordinates]
+            @incident_edges[he.origin.coordinates] = []
+          end
+          @incident_edges[he.origin.coordinates] << he
+        end
+
+        # Links all half-edges where possible.
+        # Defines +next+ and +prev+ for every half-edge by rotating
+        # through all half-edges originating at a vertex.
+        #
+        # Assuming half-edges are sorted CCW, every sequential pair of
+        # half-edges (e1, e2) can be linked by saying e1.prev = e2.twin
+        # and e2.twin.next = e1.
+        def link_half_edges
+          incident_edges.each_value do |hedges|
+            hedges.sort!
+            if hedges.size > 1
+              (0..hedges.size - 2).each do |i|
+                e1 = hedges[i]
+                e2 = hedges[i + 1]
+
+                e1.prev = e2.twin
+                e2.twin.next = e1
+              end
+
+              hedges[-1].prev = hedges[0].twin
+              hedges[0].twin.next = hedges[-1]
+
+            # handle case of dangling line
+            else
+              e = hedges[0]
+              e.twin.next = e
+              e.prev = e.twin
+            end
```

Comment:
It seems like it is the same as:
```suggestion
            # Note that for a dangling line (single edge), first edge will be its own twin.
            [*hedges, hedges.first].each_cons(2) do |e1, e2|
              e1.prev = e2.twin
              e2.twin.next = e1
            end
```

Reply:
Nice! This works